[WebGPU] Give BindGroupLayouts a notion of external textures (like BindGroups have)

litherum opened the pull request:
#### 936c42734f58a0b09f482a6ba86ba339cfa71164
<pre>
[WebGPU] Give BindGroupLayouts a notion of external textures (like BindGroups have)
<a href="https://bugs.webkit.org/show_bug.cgi?id=257055">https://bugs.webkit.org/show_bug.cgi?id=257055</a>
rdar://109585492

Reviewed by Mike Wyrzykowski.

Just like we added WGPUBindGroupExternalTextureEntry to let bind groups be able to hold
external textures, this patch adds the corresponding WGPUExternalTextureBindGroupLayoutEntry
to bind group layouts.

Note that the external texture handling added by this patch doesn&apos;t actually hook up all
the necessary pieces to the MTLArgumetnDescriptors. This patch instead just creates the
concept of a BGL&apos;s external texture, and the details can be filled in in a subsequent patch.

I&apos;ve also included a few drive-by style improvements, too.

* Source/WebCore/Modules/WebGPU/GPUExternalTextureBindingLayout.idl:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createBindGroupLayout):
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::configure):
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
(WebGPU::BindGroupLayout::entries const):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::createArgumentDescriptor):
(WebGPU::BindGroupLayout::isPresent):
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::BindGroupLayout):
(WebGPU::BindGroupLayout::entries const): Deleted.
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Pipeline.h:
* Source/WebGPU/WebGPU/PresentationContext.h:
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::convertBindingLayout):
(WebGPU::ShaderModule::convertPipelineLayout):
(WebGPU::populateBindingMember): Deleted.
* Source/WebGPU/WebGPU/WebGPUExt.h:

Canonical link: <a href="https://commits.webkit.org/264437@main">https://commits.webkit.org/264437@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/111e626cf1d323d1852c25bd94f43e01aef0222e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7553 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7821 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8004 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9193 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7742 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7564 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9777 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7745 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10618 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7684 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8795 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6971 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9301 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6103 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6874 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14574 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6995 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10347 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7489 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6114 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6830 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1812 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11038 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7223 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WebGPU/BindGroupLayout.mm:
```diff
@@ -141,13 +140,34 @@
+    if (externalTexture.nextInChain)
+        return nil;
+
+    auto descriptor = [MTLArgumentDescriptor new];
+    descriptor.dataType = MTLDataTypeTexture;
+#if USE(METAL_ARGUMENT_ACCESS_ENUMS)
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    descriptor.access = MTLArgumentAccessReadOnly;
+ALLOW_DEPRECATED_DECLARATIONS_END
+#else
+    descriptor.access = MTLBindingAccessReadOnly;
+#endif
+    return descriptor;
```
This is wrong - we'll need to add tons of stuff to the argument encoder here, not just a single texture. I think that should happen in another patch, though, so I think the only thing to change immediately here is to add a FIXME.

mwyrzykowski commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -157,28 +160,37 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
             if (!bindGroupLayout.bindingContainsStage(entry.binding, stage))
                 continue;
 
+            // FIXME: This "index" logic is fairly bogus.
```
Can we please add a bug link tracking this?

mwyrzykowski commented on Source/WebGPU/WebGPU/BindGroupLayout.mm:
```diff
@@ -141,13 +140,36 @@
+    return true;
+}
+
+static MTLArgumentDescriptor *createArgumentDescriptor(const WGPUExternalTextureBindingLayout& externalTexture)
```
I don't think this function is necessary. It duplicates `static MTLArgumentDescriptor *createArgumentDescriptor(const WGPUTextureBindingLayout& texture)`

mwyrzykowski replied:
We can certainly leave this for a second patch, but this could be implemented by returning an array of MTLArgumentDescriptor instances and using `append` instead of `unchckedAppend` when adding the descriptors

litherum replied:
Right, I created it for symmetry; there is a `createArgumentDescriptor()` for each `WGPU*BindingLayout` type. The fact that it matches right now is temporary; as the comment describes, the content of this function is totally bogus right now and will need to be fully implemented in a follow-up patch. So I think the function should still exist; it's just an unlucky accident that it does the same thing as the `WGPUTextureBindingLayout` version right now.

You're right that returning a single `MTLArgumentDescriptor` ain't gonna work when we implement this properly.

mwyrzykowski commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
```diff
@@ -175,9 +175,23 @@ Ref<BindGroupLayout> DeviceImpl::createBindGroupLayout(const BindGroupLayoutDesc
-    auto backingEntries = descriptor.entries.map([this] (const auto& entry) {
-        return WGPUBindGroupLayoutEntry {
-            nullptr,
+    auto backingExternalTextureEntries = descriptor.entries.map([] (const auto&) {
```
It's not really necessary to iterate over descriptors twice right?

Previously, there was a single call to `map`. This change keeps the call to `map` and also iterates over the data a second time.

It seems like it would be a simpler change to just store the chained entries into a vector if an external entry is encountered.

litherum replied:
I think it has to be 2 different iterations, because there are pointers from `backingEntries` into `backingExternalTextureEntries`. So, if there are any modifications to `backingExternalTextureEntries`, the vector's backing store could be reallocated, and the pointers will be dangling. So I don't think we can build them in tandem; I think we need to build the 2 vectors one after the other.

mwyrzykowski replied:
Oh great point, actually we have that problem in `Ref<BindGroup> DeviceImpl::createBindGroup(const BindGroupDescriptor& descriptor)` for which I just filed https://bugs.webkit.org/show_bug.cgi?id=257192

I suppose a container which does not re-allocate the underlying storage could be used given the size of this container is often 0 or 1, but using Vector seems fine here

litherum replied:
Yep. I'll make a follow-up patch to migrate all the places where we use this pattern to SegmentedVector.

mwyrzykowski commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -157,28 +160,37 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
         [argumentEncoder[stage] setArgumentBuffer:argumentBuffer[stage] offset:0];
     }
 
+    // FIXME: The bind group layout determines the layout of what gets put into the bind group.
```
If we iterate over the bind group layout, then I don't think we need the comments on lines 190-193, they are essentially referring to the same issue.

In either case, can we please add a bugzilla link here?

mwyrzykowski commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -157,28 +160,37 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
             if (entry.nextInChain->sType != static_cast<WGPUSType>(WGPUSTypeExtended_BindGroupEntryExternalTexture))
                 return BindGroup::createInvalid(*this);
+            if (entry.nextInChain->next)
+                return BindGroup::createInvalid(*this);
```
I think we are supposed to call `generateAValidationError("....")` in the calls to `createInvalid` according to https://www.w3.org/TR/webgpu/#dom-gpudevice-createbindgroup

Since that would be a small addition, should we please add it to this patch?

mwyrzykowski commented on Source/WebGPU/WebGPU/BindGroupLayout.mm:
```diff
@@ -141,13 +140,36 @@
+    auto descriptor = [MTLArgumentDescriptor new];
+    descriptor.dataType = MTLDataTypeTexture;
+#if USE(METAL_ARGUMENT_ACCESS_ENUMS)
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
```
`MTLBindingAccessReadOnly` has been in the SDK for some time now. Do we really still need the `ALLOW_DEPRECATED_DECLARATIONS_BEGIN / END` here if we were to change `USE_METAL_ARGUMENT_ACCESS_ENUMS` from:
```
#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MAX_ALLOWED < 150000) \
    || (PLATFORM(IOS_FAMILY) && __IPHONE_OS_VERSION_MAX_ALLOWED < 180000) \
    || (PLATFORM(APPLETV) && __TV_OS_VERSION_MAX_ALLOWED < 180000) \
    || (PLATFORM(WATCHOS) && __WATCH_OS_VERSION_MAX_ALLOWED < 110000)
#define USE_METAL_ARGUMENT_ACCESS_ENUMS 1
#endif
```
to:
```
#if (PLATFORM(MAC) && __MAC_OS_X_VERSION_MAX_ALLOWED < 140000) \
    || (PLATFORM(IOS_FAMILY) && __IPHONE_OS_VERSION_MAX_ALLOWED < 170000) \
    || (PLATFORM(APPLETV) && __TV_OS_VERSION_MAX_ALLOWED < 170000) \
    || (PLATFORM(WATCHOS) && __WATCH_OS_VERSION_MAX_ALLOWED < 100000)
```

litherum replied:
Good idea. I'd like to do this in a follow-up patch, though.

litherum replied:
https://github.com/WebKit/WebKit/pull/14235

mwyrzykowski commented on Source/WebGPU/WebGPU/BindGroupLayout.mm:
```diff
@@ -173,24 +195,47 @@ static bool containsStage(WGPUShaderStageFlags stageBitfield, auto stage)
+            if (descriptor)
+                return false;
+            descriptor = createArgumentDescriptor(type);
+            if (!descriptor)
```
`!descriptor` is already checked on line 195, so it is duplicated with this additional code line here.

I don't think it is sufficient to remove it from line 195 however, since none of the `if` conditionals may be true unless the caller guarantees that

litherum replied:
I think both are necessary.

As you say, the one on line 195 is necessary in case none of the `if`s are taken.

The one here is necessary to make sure that, if we try to create an argument descriptor of one particular type, and we fail, then we shouldn't even try to create an argument descriptor of another type. A single failure should tank the whole thing, rather than saying "well one entry is invalid but another one was valid so let's use the valid one"

mwyrzykowski replied:
That's a great point, leaving both sounds better

mwyrzykowski reviewed (approved):
Some small comments but overall looks good and merge conflict to resolve

litherum committed:
[WebGPU] Give BindGroupLayouts a notion of external textures (like BindGroups have)
https://bugs.webkit.org/show_bug.cgi?id=257055
rdar://109585492

Reviewed by Mike Wyrzykowski.

Just like we added WGPUBindGroupExternalTextureEntry to let bind groups be able to hold
external textures, this patch adds the corresponding WGPUExternalTextureBindGroupLayoutEntry
to bind group layouts.

Note that the external texture handling added by this patch doesn't actually hook up all
the necessary pieces to the MTLArgumetnDescriptors. This patch instead just creates the
concept of a BGL's external texture, and the details can be filled in in a subsequent patch.

I've also included a few drive-by style improvements, too.

* Source/WebCore/Modules/WebGPU/GPUExternalTextureBindingLayout.idl:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createBindGroupLayout):
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::configure):
* Source/WebGPU/WGSL/WGSL.h:
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
(WebGPU::BindGroupLayout::entries const):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::createArgumentDescriptor):
(WebGPU::BindGroupLayout::isPresent):
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::BindGroupLayout):
(WebGPU::BindGroupLayout::entries const): Deleted.
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Pipeline.h:
* Source/WebGPU/WebGPU/PresentationContext.h:
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::convertBindingLayout):
(WebGPU::ShaderModule::convertPipelineLayout):
(WebGPU::populateBindingMember): Deleted.
* Source/WebGPU/WebGPU/WebGPUExt.h:

Canonical link: https://commits.webkit.org/264437@main
